File: NeuralNetwork/LearningSet.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace NeuralNetwork
{
	public class LearningSet
	{
		public LearningSet(IEnumerable<Pattern> trainingData, IEnumerable<Pattern> testData, int row, int column, int classCount)
		{
			TrainingData = trainingData.ToArray();
			TestData = testData.ToArray();
			Row = row;
			Column = column;
			ClassCount = classCount;
		}
		public readonly IReadOnlyList<Pattern> TrainingData;
		public readonly IReadOnlyList<Pattern> TestData;
		public readonly int Row;
		public readonly int Column;
		public readonly int ClassCount;
		public LearningSet Subset(int trainingDataCount, int testDataCount) { return new LearningSet(TrainingData.Take(trainingDataCount), TestData.Take(testDataCount), Row, Column, ClassCount); }
	}

	public class Pattern
	{
		public Pattern(int label, double[] image)
		{
			Label = label;
			Image = image;
		}
		public readonly int Label;
		public readonly double[] Image;
		public override string ToString() { return "Image: " + Image + ", Label: " + Label; }
	}

	public static class MnistSet
	{
		static uint BigEndianToLittleEndian(uint value)
		{
			return
				(value & 0x000000FF) << 24 |
				(value & 0x0000FF00) << 8 |
				(value & 0x00FF0000) >> 8 |
				(value & 0xFF000000) >> 24;
		}
		
		public static LearningSet Load(string directoryName)
		{
			var trainSet = LoadInternal(directoryName + "/train");
			var testSet = LoadInternal(directoryName + "/t10k");
			if (trainSet.Item2 != testSet.Item2 || trainSet.Item3 != testSet.Item3)
				return null;
			return new LearningSet(trainSet.Item1, testSet.Item1, trainSet.Item2, trainSet.Item3, 10);
		}

		static Tuple<Pattern[], int, int> LoadInternal(string fileName)
		{
			using (FileStream labelFile = new FileStream(fileName + "-labels.idx1-ubyte", FileMode.Open))
			using (BinaryReader labelReader = new BinaryReader(labelFile))
			using (FileStream imageFile = new FileStream(fileName + "-images.idx3-ubyte", FileMode.Open))
			using (BinaryReader imageReader = new BinaryReader(imageFile))
			{
				if (BigEndianToLittleEndian(labelReader.ReadUInt32()) != 0x801)
					return null;
				if (BigEndianToLittleEndian(imageReader.ReadUInt32()) != 0x803)
					return null;
				var length = (int)BigEndianToLittleEndian(labelReader.ReadUInt32());
				if (length != BigEndianToLittleEndian(imageReader.ReadUInt32()))
					return null;
				var row = (int)BigEndianToLittleEndian(imageReader.ReadUInt32());
				var column = (int)BigEndianToLittleEndian(imageReader.ReadUInt32());
				Pattern[] data = new Pattern[length];
				for (int i = 0; i < data.Length; i++)
				{
					var label = labelReader.ReadByte();
					var image = new double[row * column];
					for (int j = 0; j < image.Length; j++)
						image[j] = (double)imageReader.ReadByte() / byte.MaxValue;
					data[i] = new Pattern(label, image);
				}
				return new Tuple<Pattern[], int, int>(data, row, column);
			}
		}
	}

	public static class PatternRecognitionSet
	{
		public static LearningSet Load(string directoryName)
		{
			return new LearningSet(
				LoadPatterns(directoryName + "/pattern2learn.dat"),
				LoadPatterns(directoryName + "/pattern2recog.dat"),
				7, 5, 10
			);
		}

		static IEnumerable<Pattern> LoadPatterns(string fileName)
		{
			using (StreamReader reader = new StreamReader(fileName))
			{
				string line;
				while ((line = reader.ReadLine()) != null)
				{
					var elems = line.Split(new[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries);
					yield return new Pattern(int.Parse(elems[0]), elems.Skip(1).Take(elems.Length - 1).Select(x => double.Parse(x)).ToArray());
				}
			}
		}
	}
}

File: NativeNeuralNetwork/LearningSet.cpp
#include "LearningSet.h"

template <class T> inline T* pointer_cast(void* pointer) { return static_cast<T*>(pointer); }

DataSet& DataSet::operator=(DataSet&& dataset)
{
	if (this != &dataset)
	{
		labels = std::move(dataset.labels);
		images = std::move(dataset.images);
		row = dataset.row;
		column = dataset.column;

		dataset.row = 0;
		dataset.column = 0;
	}
	return *this;
}

void DataSet::CopyFrom(const DataSet& dataset, size_t count)
{
	assert(count > 0 && count <= dataset.Count());
	Allocate(count, dataset.row, dataset.column);
	for (unsigned int i = 0; i < count; i++)
	{
		labels[i] = dataset.labels[i];
		images[i] = dataset.images[i];
	}
}

void DataSet::Allocate(size_t length, unsigned int newRow, unsigned int newColumn)
{
	assert(newRow * newColumn > 0);
	labels.resize(length);
	images.resize(length);
	for (size_t i = 0; i < length; i++)
		images[i].resize(newRow * newColumn);
	row = newRow;
	column = newColumn;
}

inline uint32_t ReadInt32BigEndian(std::ifstream& stream)
{
	uint32_t temp;
	stream.read(pointer_cast<char>(&temp), sizeof(temp));
	return
		(temp & 0x000000FF) << 24 |
		(temp & 0x0000FF00) << 8 |
		(temp & 0x00FF0000) >> 8 |
		(temp & 0xFF000000) >> 24;
}

inline uint8_t ReadByte(std::ifstream& stream)
{
	uint8_t temp;
	stream.read(pointer_cast<char>(&temp), sizeof(temp));
	return temp;
}

void LoadMnistSetInternal(DataSet& dataset, const std::string& fileName)
{
	std::ifstream labelFile((fileName + "-labels.idx1-ubyte").c_str(), std::ios::binary | std::ios::in);
	std::ifstream imageFile((fileName + "-images.idx3-ubyte").c_str(), std::ios::binary | std::ios::in);
	if (ReadInt32BigEndian(labelFile) != 0x801)
		return;
	if (ReadInt32BigEndian(imageFile) != 0x803)
		return;
	auto length = ReadInt32BigEndian(labelFile);
	if (length != ReadInt32BigEndian(imageFile))
		return;
	auto row = ReadInt32BigEndian(imageFile);
	auto column = ReadInt32BigEndian(imageFile);
	auto imageLength = row * column;
	dataset.Allocate(length, row, column);
	for (uint32_t i = 0; i < length; i++)
	{
		dataset.Labels()[i] = ReadByte(labelFile);
		for (uint32_t j = 0; j < imageLength; j++)
			dataset.Images()[i][j] = static_cast<ValueType>(ReadByte(imageFile)) / std::numeric_limits<unsigned char>::max();
	}
}

LearningSet LoadMnistSet(const std::string& directoryName)
{
	LearningSet set;
	LoadMnistSetInternal(set.TrainingData(), directoryName + "/train");
	LoadMnistSetInternal(set.TestData(), directoryName + "/t10k");
	set.ClassCount = 10;
	return set;
}
